File: skyspark-scala/src/test/scala/com/akikanellis/skyspark/scala/api/algorithms/bnl/FlagProducerTest.scala
package com.akikanellis.skyspark.scala.api.algorithms.bnl

import com.akikanellis.skyspark.scala.api.algorithms.Point
import com.akikanellis.skyspark.scala.test_utils.UnitSpec

class FlagProducerTest extends UnitSpec {
  var median: Point = _
  var flagProducer: FlagProducer = _

  before {
    median = Point(5, 5)
    flagProducer = new FlagProducer(median)
  }

  "A point with x and y smaller than median" should "produce flag 00" in {
    val point = Point(4, 4)
    val expectedFlag = Flag(false, false)

    val actualFlag = flagProducer.calculateFlag(point)

    actualFlag shouldBe expectedFlag
  }

  "A point with x and y bigger than median" should "produce flag 11" in {
    val point = Point(6, 6)
    val expectedFlag = Flag(true, true)

    val actualFlag = flagProducer.calculateFlag(point)

    actualFlag shouldBe expectedFlag
  }

  "A point with x and y equal to median" should "produce flag 11" in {
    val point = Point(5, 5)
    val expectedFlag = Flag(true, true)

    val actualFlag = flagProducer.calculateFlag(point)

    actualFlag shouldBe expectedFlag
  }

  "A point with x bigger and y smaller than median" should "produce flag 10" in {
    val point = Point(6, 4)
    val expectedFlag = Flag(true, false)

    val actualFlag = flagProducer.calculateFlag(point)

    actualFlag shouldBe expectedFlag
  }

  "A point with x smaller and y bigger than median" should "produce flag 01" in {
    val point = Point(4, 6)
    val expectedFlag = Flag(false, true)

    val actualFlag = flagProducer.calculateFlag(point)

    actualFlag shouldBe expectedFlag
  }
}

File: skyspark-java/src/test/java/com/akikanellis/skyspark/api/algorithms/bnl/FlagProducerTest.java
package com.akikanellis.skyspark.api.algorithms.bnl;

import com.akikanellis.skyspark.api.algorithms.Point;
import com.akikanellis.skyspark.api.test_utils.assertions.MoreAssertions;
import org.junit.Before;
import org.junit.Test;

public class FlagProducerTest {
    private FlagProducer flagProducer;

    @Before public void beforeEach() {
        Point median = new Point(5, 5);
        flagProducer = new FlagProducer(median);
    }

    @Test public void point_withDimensionsSmallerThanMedian_producesFlag00() {
        Point point = new Point(4, 4);
        Flag expectedFlag = new Flag(false, false);

        assertFlagIsAsExpected(point, expectedFlag);
    }

    @Test public void point_withDimensionsLargerThanMedian_producesFlag11() {
        Point point = new Point(6, 6);
        Flag expectedFlag = new Flag(true, true);

        assertFlagIsAsExpected(point, expectedFlag);
    }

    @Test public void point_withDimensionsEqualToMedian_producesFlag11() {
        Point point = new Point(5, 5);
        Flag expectedFlag = new Flag(true, true);

        assertFlagIsAsExpected(point, expectedFlag);
    }

    @Test public void point_withFirstDimensionLargerAndSecondSmaller_producesFlag10() {
        Point point = new Point(6, 4);
        Flag expectedFlag = new Flag(true, false);

        assertFlagIsAsExpected(point, expectedFlag);
    }

    @Test public void point_withFirstDimensionSmallerAndSecondLarger_producesFlag01() {
        Point point = new Point(4, 6);
        Flag expectedFlag = new Flag(false, true);

        assertFlagIsAsExpected(point, expectedFlag);
    }

    private void assertFlagIsAsExpected(Point point, Flag expectedFlag) {
        Flag actualFlag = flagProducer.calculateFlag(point);

        MoreAssertions.assertThat(actualFlag).isEqualTo(expectedFlag);
    }
}